soc/st/stm32/common/stm32_cache.c

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2025 Henrik Lindblom <henrik.lindblom@vaisala.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "stm32_cache.h"
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/mem_mgmt/mem_attr.h>
+#include <zephyr/dt-bindings/memory-attr/memory-attr-arm.h>
+
+/* Returns true if given buffer is in a non-cacheable SRAM region.
+ * This is determined using the device tree, meaning the .nocache region won't work.
+ * The entire buffer must be in a single region.
+ * An example of how the SRAM region can be defined in the DTS:
+ *	&sram4 {
+ *		zephyr,memory-attr = <( DT_MEM_ARM(ATTR_MPU_RAM_NOCACHE) | ... )>;
+ *	};
+ */
+bool buf_in_nocache(uintptr_t buf, size_t len_bytes)
+{
+#ifndef CONFIG_DCACHE
+	ARG_UNUSED(buf);
+	ARG_UNUSED(len_bytes);
+	return false;
+#else
+	bool buf_within_nocache = false;
+
+#ifdef CONFIG_EXTERNAL_CACHE
+	buf_within_nocache = (buf >= ((uintptr_t)_image_ram_start) &&
+			      (buf + len_bytes - 1) <= ((uintptr_t)_image_ram_end));
+	if (buf_within_nocache) {
+		return true;
+	}
+#endif /* CONFIG_EXTERNAL_CACHE */
+
+#ifdef CONFIG_NOCACHE_MEMORY
+	buf_within_nocache = (buf >= ((uintptr_t)_nocache_ram_start)) &&
+			     ((buf + len_bytes - 1) <= ((uintptr_t)_nocache_ram_end));
+	if (buf_within_nocache) {
+		return true;
+	}
+#endif /* CONFIG_NOCACHE_MEMORY */
+
+#ifdef CONFIG_MEM_ATTR
+	buf_within_nocache =
+		mem_attr_check_buf((void *)buf, len_bytes, DT_MEM_ARM_MPU_RAM_NOCACHE) == 0;
+	if (buf_within_nocache) {
+		return true;
+	}
```

Comment:
Ditto (+ bring back comment)
```suggestion
	/* Check if buffer is in a region marked NOCACHE in DT */
	if (mem_attr_check_buf((void *)buf, len_bytes, DT_MEM_ARM_MPU_RAM_NOCACHE) == 0) {
		return true;
	}
```